iroha/src/block.rs

```diff
@@ -25,6 +26,47 @@ use crate::{
     tx::{VersionedAcceptedTransaction, VersionedValidTransaction},
 };
 
+/// Blockchain.
+#[derive(Debug, Default)]
+pub struct Chain {
+    blocks: DashMap<u64, VersionedCommittedBlock>,
```

Comment:
I guess a better structure would be:
```
pub struct Chain {
    blocks: DashMap<Hash, VersionedCommittedBlock>,
    first: Hash,
    last: Hash,
}
```

As it is right now doesn't seem to be promissing

Reply:
Why do you think so? This is our internal representation where we can garantee that they are connected after validation, and we mainly want to retrieve them in ordered sequence - similar to vec and its access by position before.

Reply:
(Changed code a bit)

I mean it just looks kinda ugly. It is an array which is internally is dashmap. Even though you would likely to reference some chain part, rather than its index, so it is better to have Hash as key.

Reply:
Yes, it is currently used as an array, and with dashmap it is more logical to use it this way.

Reply:
it is more logical to use it like array? I think no, indeces are meaningless except for first and last

Reply:
I tried this way, then I remembered that we can not mutate self. The `Chain` struct in all its methods takes immutable reference. And we can't use RwLock here (sync - because it blocks, async - because we do not want this part to be async because of the WSV usage). So this needs something like atomic option, but I don't think it is needed enough to introduce it.

Reply:
You could use atomic from crossbeam. It has nice api and will be lockfree for hash, as it is just an array inside. (we can even do static compile check for that)

Reply:
Let's try this in a different PR. We still have to check if the crossbeam is already in our deps tree.

Reply:
There is a problem that the push method should be atomic, but when we will update two atomics inside it already will not be atomic and something can happen in between.

Reply:
That makes sence. May be we can add a spinrwlock for that? (I mean, we know that adding will be a short path, so it will be safe and lockfree)

Reply:
I think it is over-engineering in this case. It is better to do this when we need the structure that you outlined.